File: ProjectTracker20vb/ProjectTracker.Library/ResourceAssignment.vb
<Serializable()> _
Public Class ResourceAssignment
  Inherits BusinessBase(Of ResourceAssignment)

  Implements IHoldRoles

#Region " Business Methods "

  Private mProjectId As Guid = Guid.Empty
  Private mProjectName As String = ""
  Private mAssigned As New SmartDate(Today)
  Private mRole As Integer
  Private mTimestamp(7) As Byte

  Public ReadOnly Property ProjectId() As Guid
    Get
      CanReadProperty(True)
      Return mProjectId
    End Get
  End Property

  Public ReadOnly Property ProjectName() As String
    Get
      CanReadProperty(True)
      Return mProjectName
    End Get
  End Property

  Public ReadOnly Property Assigned() As String
    Get
      CanReadProperty(True)
      Return mAssigned.Text
    End Get
  End Property

  Public Property Role() As Integer Implements IHoldRoles.Role
    Get
      CanReadProperty(True)
      Return mRole
    End Get
    Set(ByVal value As Integer)
      CanWriteProperty(True)
      If Not mRole.Equals(value) Then
        mRole = value
        PropertyHasChanged()
      End If
    End Set
  End Property

  Public Function GetProject() As Project

    Return Project.GetProject(mProjectId)

  End Function

  Protected Overrides Function GetIdValue() As Object

    Return mProjectId

  End Function

#End Region

#Region " Validation Rules "

  Protected Overrides Sub AddBusinessRules()

    ValidationRules.AddRule(AddressOf Assignment.ValidRole, "Role")

  End Sub

#End Region

#Region " Authorization Rules "

  Protected Overrides Sub AddAuthorizationRules()

    AuthorizationRules.AllowWrite("Role", "ProjectManager")

  End Sub

#End Region

#Region " Factory Methods "

  Friend Shared Function NewResourceAssignment( _
    ByVal projectId As Guid) As ResourceAssignment

    Return New ResourceAssignment(Project.GetProject(projectId), RoleList.DefaultRole)

  End Function

  Friend Shared Function GetResourceAssignment( _
    ByVal dr As SafeDataReader) As ResourceAssignment

    Return New ResourceAssignment(dr)

  End Function

  Private Sub New()

    MarkAsChild()

  End Sub

  Private Sub New(ByVal dr As SafeDataReader)

    MarkAsChild()
    Fetch(dr)

  End Sub

  Private Sub New(ByVal project As Project, ByVal role As Integer)

    MarkAsChild()
    mProjectId = project.Id
    mProjectName = project.Name
    mAssigned.Date = Assignment.GetDefaultAssignedDate
    mRole = role

  End Sub

#End Region

#Region " Data Access "

  Private Sub Fetch(ByVal dr As SafeDataReader)

    With dr
      mProjectId = .GetGuid("ProjectId")
      mProjectName = .GetString("Name")
      mAssigned = .GetSmartDate("Assigned")
      mRole = .GetInt32("Role")
      .GetBytes("LastChanged", 0, mTimestamp, 0, 8)
    End With
    MarkOld()

  End Sub

  Friend Sub Insert(ByVal cn As SqlConnection, ByVal resource As Resource)

    ' if we're not dirty then don't update the database
    If Not Me.IsDirty Then Exit Sub

    mTimestamp = Assignment.AddAssignment( _
      cn, mProjectId, resource.Id, mAssigned, mRole)
    MarkOld()

  End Sub

  Friend Sub Update( _
    ByVal cn As SqlClient.SqlConnection, ByVal resource As Resource)

    ' if we're not dirty then don't update the database
    If Not Me.IsDirty Then Exit Sub

    mTimestamp = Assignment.UpdateAssignment( _
      cn, mProjectId, resource.Id, mAssigned, mRole, mTimestamp)
    MarkOld()

  End Sub

  Friend Sub DeleteSelf(ByVal cn As SqlClient.SqlConnection, ByVal resource As Resource)

    ' if we're not dirty then don't update the database
    If Not Me.IsDirty Then Exit Sub

    ' if we're new then don't update the database
    If Me.IsNew Then Exit Sub

    Assignment.RemoveAssignment(cn, mProjectId, resource.Id)
    MarkNew()

  End Sub

#End Region

End Class

File: ProjectTracker20cs/ProjectTracker.Library/ResourceAssignment.cs
using System;
using System.Data;
using System.Data.SqlClient;
using Csla;
using Csla.Data;

namespace ProjectTracker.Library
{
  [Serializable()]
  public class ResourceAssignment : BusinessBase<ResourceAssignment>, IHoldRoles
  {
    #region Business Methods

    private Guid _projectId = Guid.Empty;
    private string _projectName = string.Empty;
    private SmartDate _assigned = new SmartDate(DateTime.Today);
    private int _role;
    private byte[] _timestamp = new byte[8];

    public Guid ProjectId
    {
      [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.NoInlining)]
      get
      {
        CanReadProperty(true);
        return _projectId;
      }
    }

    public string ProjectName
    {
      [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.NoInlining)]
      get
      {
        CanReadProperty(true);
        return _projectName;
      }
    }

    public string Assigned
    {
      [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.NoInlining)]
      get
      {
        CanReadProperty(true);
        return _assigned.Text;
      }
    }

    public int Role
    {
      [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.NoInlining)]
      get
      {
        CanReadProperty(true);
        return _role;
      }
      [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.NoInlining)]
      set
      {
        CanWriteProperty(true);
        if (!_role.Equals(value))
        {
          _role = value;
          PropertyHasChanged();
        }
      }
    }

    public Project GetProject()
    {
      return Project.GetProject(_projectId);
    }

    protected override object GetIdValue()
    {
      return _projectId;
    }

    #endregion

    #region Validation Rules

    protected override void AddBusinessRules()
    {
      ValidationRules.AddRule(
        new Csla.Validation.RuleHandler(
          Assignment.ValidRole), "Role");
    }

    #endregion

    #region Authorization Rules

    protected override void AddAuthorizationRules()
    {
      AuthorizationRules.AllowWrite(
        "Role", "ProjectManager");
    }

    #endregion

    #region Factory Methods

    internal static ResourceAssignment NewResourceAssignment(
      Guid projectId)
    {
      return new ResourceAssignment(
        Project.GetProject(projectId), RoleList.DefaultRole());
    }

    internal static ResourceAssignment GetResourceAssignment(
      SafeDataReader dr)
    {
      return new ResourceAssignment(dr);
    }

    private ResourceAssignment()
    { MarkAsChild(); }

    #endregion

    #region Data Access

    /// <summary>
    /// Called to when a new object is created.
    /// </summary>
    private ResourceAssignment(Project project, int role)
    {
      MarkAsChild();
      _projectId = project.Id;
      _projectName = project.Name;
      _assigned.Date = Assignment.GetDefaultAssignedDate();
      _role = role;
    }

    /// <summary>
    /// Called when loading data from the database.
    /// </summary>
    private ResourceAssignment(SafeDataReader dr)
    {
      MarkAsChild();
      _projectId = dr.GetGuid("ProjectId");
      _projectName = dr.GetString("Name");
      _assigned = dr.GetSmartDate("Assigned");
      _role = dr.GetInt32("Role");
      dr.GetBytes("LastChanged", 0, _timestamp, 0, 8);
      MarkOld();
    }

    internal void Insert(SqlConnection cn, Resource resource)
    {
      // if we're not dirty then don't update the database
      if (!this.IsDirty) return;

      _timestamp = Assignment.AddAssignment(
        cn, _projectId, resource.Id, _assigned, _role);
      MarkOld();
    }

    internal void Update(SqlConnection cn, Resource resource)
    {
      // if we're not dirty then don't update the database
      if (!this.IsDirty) return;

      _timestamp = Assignment.UpdateAssignment(
        cn, _projectId, resource.Id, _assigned, _role, _timestamp);
      MarkOld();
    }

    internal void DeleteSelf(SqlConnection cn, Resource resource)
    {
      // if we're not dirty then don't update the database
      if (!this.IsDirty) return;

      // if we're new then don't update the database
      if (this.IsNew) return;

      Assignment.RemoveAssignment(cn, _projectId, resource.Id);
      MarkNew();
    }

    #endregion

  }
}
